src/cpu/flexcpu/simple_dataflow_thread.hh

```diff
@@ -75,56 +98,118 @@ class SDCPUThread
      * execution of an instruction, this should be freely usable as a tag for
      * our dynamic instructions.
      */
-    typedef uint64_t InstTag;
-    InstTag numInstsCommitted = 0;
-    InstTag numInstsSpeculated = 0;
+    InstSeqNum numInstsCommitted = 0;
 
     /**
-     * The SimpleThread class
+     * The SimpleThread class already contains all internal state necessary to
+     * hold onto the state of a basic CPU, demonstrated by its use in many
+     * other CPU models. Until we find a significant need to hold particular
+     * permanent committed state different from a regular CPU in this library,
+     * there is no reason to redefine the wheel.
+     * 
+     * The committed state represents the state resulting from the complete
+     * correct execution of all instructions up to the point represented by
+     * numInstsCommitted.
      */
     SimpleThread _committedState;
 
     // BEGIN Speculative state
 
     /**
-     * Instruction execution will pass through an ExecContext interface,
-     * therefore in most cases, it will be tied to speculative state, instead
-     * of the "real" state of the system.
+     * We hold onto the InstSeqNum of the last instruction successfully
+     * committed, for ease of checking if older instruction dependencies have
+     * already been satisfied.
      */
-    SDCPUExecContext _isaInterface;
+    InstSeqNum lastCommittedInstNum;
 
+    /**
+     * We hold a "speculative" PC ahead of the "true" committed PC, in order to
+     * track what it is we're fetching next, prior to instruction issue. The
+     * "true" PC is only advanced during instruction commit.
+     */
     TheISA::PCState _speculativePC;
+
+    InstSeqNum numInstsSpeculated = 0;
+
+    /**
+     * We hold a vector of all in-flight instructions, added during issue and
+     * removed after commit. If we always issue in order, this vector should
+     * always be sorted, in the order that the instructions should be committed
+     * so replacing this with a queue might be a better idea hmm.
+     */
+    std::vector<shared_ptr<InflightInst>> inflightInsts;
     // TODO hold speculative state.
     // TODO hold table of in-flight instructions.
 
     // END Speculative state
 
+    // BEGIN Internal functions
+
+    /**
+     * Utility function for 
+     */
+    void commitAllAvailableInstructions();
+
+    /**
+     * Utility function for committing the results of a previously in-flight
+     * instruction. This function should update the _committedState using the
+     * in-flight instruction's stored results, increment numInstsCommitted, and
+     * make no further changes to the thread state.
+     */
+    void commitInstruction(shared_ptr<InflightInst> inst);
+
     // TODO make sure this function is capable of holding off on issue in the
     // case where we've only partially read in a macroOp
```

Comment:
To begin with, use RISC-V and don't worry about this. At least, I think you can ignore this with RISC-V